src/NHibernate/Collection/Trackers/MapQueueOperationTracker.cs

```diff
@@ -0,0 +1,212 @@
+	{
+		private readonly ICollectionPersister _collectionPersister;
+		private IDictionary<TKey, TValue> _queue;
+		private IDictionary<TKey, TValue> _orphanMap;
+		private ISet<TKey> _removalQueue;
+		private int _queueSize;
+
+		public MapQueueOperationTracker(ICollectionPersister collectionPersister)
+		{
+			_collectionPersister = collectionPersister;
+		}
+
+		public override void AfterFlushing()
+		{
+			_queue?.Clear();
+			_removalQueue?.Clear();
+			_queueSize = 0;
+			_orphanMap = null;
+		}
+
+		/// <inheritdoc />
+		public override void ClearCollection()
+		{
+			AfterFlushing();
+			Cleared = true;
+		}
+
+		/// <inheritdoc />
+		public override IEnumerable GetAddedElements()
+		{
+			return _queue?.Values ?? (IEnumerable) Enumerable.Empty<object>();
+		}
+
+		/// <inheritdoc />
+		public override IEnumerable GetOrphans()
+		{
+			return _orphanMap?.Values ?? (IEnumerable) Enumerable.Empty<object>();
+		}
+
+		/// <inheritdoc />
+		public override void AddElementByKey(TKey elementKey, TValue element)
+		{
+			_removalQueue?.Remove(elementKey); // We have to remove the key from the removal list when the element is readded
+			GetOrCreateQueue().Add(elementKey, element);
+			if (!Cleared)
+			{
+				_queueSize++;
+			}
+		}
+
+		/// <inheritdoc />
+		public override bool ContainsKey(TKey key)
+		{
+			return _queue?.ContainsKey(key) == true;
+		}
+
+		/// <inheritdoc />
+		public override int GetQueueSize()
+		{
+			return Cleared ? (_queue?.Count ?? 0) : _queueSize;
+		}
+
+		/// <inheritdoc />
+		public override bool IsElementKeyQueuedForDelete(TKey elementKey)
+		{
+			return _removalQueue?.Contains(elementKey) ?? false;
+		}
+
+		/// <inheritdoc />
+		public override bool RemoveElementByKey(TKey elementKey, TValue oldElement, bool? existsInDb)
+		{
+			if (Cleared)
+			{
+				return _queue?.Remove(elementKey) ?? false;
+			}
+
+			// We can have the following scenarios:
+			// 1. remove a key that exists in db and it is not in the queue and removal queue (decrease queue size)
+			// 2. remove a key that exists in db and it is in the queue (decrease queue size)
+			// 3. remove a key that does not exist in db and it is not in the queue (don't decrease queue size)
+			// 4. remove a key that does not exist in db and it is in the queue (decrease queue size)
+			// 5. remove a key that exists in db and it is in the removal queue (don't decrease queue size)
+
+			// If the key is not present in the database and in the queue, do nothing
+			if (existsInDb == false && _queue?.ContainsKey(elementKey) != true)
+			{
+				return false;
+			}
+
+			if (existsInDb == true)
+			{
+				GetOrCreateOrphanMap()[elementKey] = oldElement;
+			}
+
+			// We don't want to have non database keys in the removal queue
+			if (_queue?.Remove(elementKey) == true |
+				(_orphanMap?.ContainsKey(elementKey) == true && GetOrCreateRemovalQueue().Add(elementKey)))
```

Comment:
Why can't we just use `existsInDb == true` instead of `_orphanMap?.ContainsKey(elementKey) == true`?

Reply:
When removing an element that exists in the database twice (remove, add, remove), in the second remove `existsInDb` will be `null` because the element is in the queue, that is why we cannot rely on `existsInDb`. An example can be found in [MapRemove](https://github.com/nhibernate/nhibernate-core/pull/2010/files#diff-00e2aadba38d74d8484b0e6fb6071d90R2178) test.